File: src/btc5m/research.py
"""Descriptive strategy comparisons from existing paper accounting, never trade signals."""

from __future__ import annotations

import json
import random
import re
import statistics
import threading
import time
from collections import defaultdict
from dataclasses import asdict, dataclass
from decimal import Decimal
from pathlib import Path
from typing import Any

from btc5m.ledger import Ledger

D = Decimal
ROUND_MS = 300000
REPS = 2000


@dataclass
class RoundResult:
    slug: str
    start_ms: int
    net: Decimal
    fees: Decimal
    used: bool
    complete: bool
    observed: bool
    uncertain: bool
    traded_shares: Decimal = D(0)
    entry_price: Decimal | None = None
    entry_seconds: float | None = None
    direction: str | None = None
    exit_reason: str | None = None


def performance(rows: list[RoundResult]) -> dict[str, Any]:
    values = [r.net for r in rows]
    wins, losses = [v for v in values if v > 0], [v for v in values if v < 0]
    n, net = len(rows), sum(values, D(0))
    fees = sum((r.fees for r in rows), D(0))
    shares = sum((r.traded_shares for r in rows), D(0))
    positive = sorted(wins, reverse=True)

    def removed(k: int) -> Decimal | None:
        count = min(k, len(positive))
        return net - sum(positive[:count], D(0)) if n > count else None

    return {
        "rounds": n,
        "flagged_rounds": sum(r.uncertain for r in rows),
        "net": net,
        "fees": fees,
        "before_fees": net + fees,
        "mean": net / n if n else None,
        "median": statistics.median(values) if n else None,
        "wins": len(wins),
        "losses": len(losses),
        "flat": n - len(wins) - len(losses),
        "win_rate": len(wins) / n if n else None,
        "average_win": sum(wins, D(0)) / len(wins) if wins else None,
        "average_loss": sum(losses, D(0)) / len(losses) if losses else None,
        "profit_factor": sum(wins, D(0)) / -sum(losses, D(0)) if losses else None,
        "best": max(values) if n else None,
        "worst": min(values) if n else None,
        "best_win_share": positive[0] / sum(positive, D(0)) if positive else None,
        "without_best_1": removed(1),
        "without_best_3": removed(3),
        "cost_headroom_per_share": net / shares if shares and net > 0 else None,
        "cost_stress": [
            {"extra_per_share": extra, "extra_cost": shares * extra, "net": net - shares * extra}
            for extra in map(D, ("0", ".0025", ".005", ".01"))
        ],
    }


def usable(row: RoundResult, cohort: str) -> bool:
    return row.observed and row.complete and (cohort == "all" or not row.uncertain)


def block_sensitivity(values: dict[int, float], *, block_rounds: int) -> dict[str, Any]:
    """Resample whole, nonoverlapping UTC blocks; never connect across missing slots.

    Percentiles describe the empirical block distribution, assuming blocks represent
    future conditions. Not selection-adjusted inference or a probability of profit.
    """
    width = block_rounds * ROUND_MS
    groups: dict[int, dict[int, float]] = defaultdict(dict)
    for stamp, value in values.items():
        groups[stamp // width * width][stamp] = value
    blocks = [
        sum(group.values())
        for start, group in sorted(groups.items())
        if set(group) == set(range(start, start + width, ROUND_MS))
    ]
    n = len(blocks)
    result: dict[str, Any] = {
        "minutes": block_rounds * 5,
        "complete_blocks": n,
        "included_rounds": n * block_rounds,
        "available_rounds": len(values),
        "interval95": None,
        "mean_per_round": sum(blocks) / (n * block_rounds) if n else None,
        "status": "too_few_blocks",
        "resamples": REPS,
    }
    if n < 6:
        return result
    if all(v == 0 for v in blocks):
        result["status"] = "no_block_variation"
        return result
    rng = random.Random(6027 + block_rounds)
    means = sorted(sum(rng.choices(blocks, k=n)) / (n * block_rounds) for _ in range(REPS))
    result.update(
        interval95=[means[int(0.025 * (REPS - 1))], means[int(0.975 * (REPS - 1))]],
        status="descriptive",
    )
    return result


def _price_group(row: RoundResult) -> str:
    p = row.entry_price
    if p is None:
        return "Unknown / conversion"
    for limit, label in ((".1", "Below 10¢"), (".3", "10–30¢"), (".7", "30–70¢"), (".9", "70–90¢")):
        if p < D(limit):
            return label
    return "90¢ or more"


def _time_group(row: RoundResult) -> str:
    if row.entry_seconds is None:
        return "Unknown / conversion"
    for limit, label in (
        (60, "Below 60s"),
        (90, "60–90s"),
        (120, "90–120s"),
        (150, "120–150s"),
        (180, "150–180s"),
    ):
        if row.entry_seconds < limit:
            return label
    return "180s or more"


def summarize(rows: list[RoundResult], *, cohort: str) -> dict[str, Any]:
    chosen = [r for r in rows if r.used and usable(r, cohort)]
    groups = {}
    group_order = {
        "entry_price": [
            "Below 10¢",
            "10–30¢",
            "30–70¢",
            "70–90¢",
            "90¢ or more",
            "Unknown / conversion",
        ],
        "entry_time": [
            "Below 60s",
            "60–90s",
            "90–120s",
            "120–150s",
            "150–180s",
            "180s or more",
            "Unknown / conversion",
        ],
    }
    for name, group_by in (
        ("entry_price", _price_group),
        ("entry_time", _time_group),
        ("direction", lambda r: r.direction or "Unknown / conversion"),
        ("exit_reason", lambda r: r.exit_reason or "Settlement / merge"),
    ):
        partition: dict[str, list[RoundResult]] = defaultdict(list)
        for row in chosen:
            partition[group_by(row)].append(row)
        labels = group_order.get(name, sorted(partition))
        groups[name] = [
            {"label": label, **performance(partition[label])}
            for label in labels
            if label in partition
        ]
    hours: dict[int, list[RoundResult]] = defaultdict(list)
    for row in rows:
        hours[row.start_ms // 3600000 * 3600000].append(row)
    hourly = []
    cumulative = D(0)
    for stamp, sample in sorted(hours.items()):
        selected = [r for r in sample if r.used and usable(r, cohort)]
        net = sum((r.net for r in selected), D(0))
        cumulative += net
        hourly.append(
            {
                "at_ms": stamp,
                "net": net,
                "cumulative": cumulative,
                "used_rounds": len(selected),
                "excluded_used": sum(r.used for r in sample) - len(selected),
                "observed_slots": sum(r.observed for r in sample),
                "flagged_rounds": sum(r.uncertain for r in selected),
            }
        )
    # Only observable completed/flat rounds enter calendar-time comparisons. Missing
    # and still-held rounds are never treated as flat returns.
    calendar = {r.start_ms: float(r.net) for r in rows if usable(r, cohort)}
    ordered = sorted(chosen, key=lambda r: (-r.net, r.start_ms))
    return {
        "stats": performance(chosen),
        "counts": {
            "observed": sum(r.observed for r in rows),
            "used": sum(r.used for r in rows),
            "flagged_used": sum(r.used and r.uncertain for r in rows),
            "incomplete_used": sum(r.used and not r.complete for r in rows),
            "flat_observed": sum(not r.used and r.observed and r.complete for r in rows),
        },
        "hourly": hourly,
        "groups": groups,
        "block_sensitivity": [block_sensitivity(calendar, block_rounds=b) for b in (6, 12)],
        "best_rounds": [asdict(r) for r in ordered[:5]],
        "worst_rounds": [asdict(r) for r in sorted(chosen, key=lambda r: (r.net, r.start_ms))[:5]],
        "rounds": [asdict(r) for r in chosen],
    }


def compare_rounds(a: list[RoundResult], b: list[RoundResult], *, cohort: str) -> dict[str, Any]:
    left, right = {r.slug: r for r in a}, {r.slug: r for r in b}
    matched = [
        (left[k], right[k])
        for k in sorted(left.keys() & right.keys())
        if usable(left[k], cohort) and usable(right[k], cohort)
    ]
    differences = {x.start_ms: float(x.net - y.net) for x, y in matched}
    active = [(x, y) for x, y in matched if x.used or y.used]
    both = sum(x.used and y.used for x, y in matched)
    xnet, ynet = sum((x.net for x, y in matched), D(0)), sum((y.net for x, y in matched), D(0))
    correlation = None
    if len(active) >= 3:
        xs, ys = [float(x.net) for x, y in active], [float(y.net) for x, y in active]
        if statistics.pvariance(xs) > 0 and statistics.pvariance(ys) > 0:
            correlation = statistics.correlation(xs, ys)
    return {
        "matched_rounds": len(matched),
        "excluded_rounds": len(left.keys() | right.keys()) - len(matched),
        "both_used": both,
        "both_flat": len(matched) - len(active),
        "selected_only": sum(x.used and not y.used for x, y in matched),
        "comparison_only": sum(y.used and not x.used for x, y in matched),
        "selected_net": xnet,
        "comparison_net": ynet,
        "difference": xnet - ynet,
        "common_used_fraction": both / len(active) if active else None,
        "active_round_correlation": correlation,
        "block_sensitivity": [block_sensitivity(differences, block_rounds=k) for k in (6, 12)],
        "series": [
            {
                "at_ms": x.start_ms,
                "selected": x.net,
                "comparison": y.net,
                "difference": x.net - y.net,
            }
            for x, y in matched
        ],
    }


def read_rounds(
    path: Path, ident: str, mode: str, phase: dict[str, Any], tape_id: str
) -> tuple[list[RoundResult], dict[str, int]]:
    ledger = Ledger(path, "0x" + ident.zfill(40), readonly=True, environment="paper")
    try:
        if ledger.environment != "paper":
            raise ValueError("PAPER_RESEARCH_ONLY")
        ledger.db.execute("BEGIN")
        identity = json.loads(ledger._meta("lab_identity") or "{}")
        if identity != {
            "variant": ident,
            "tape": tape_id,
            "start_ms": phase["start_ms"],
            "end_ms": phase.get("end_ms"),
        }:
            raise ValueError("RESEARCH_IDENTITY_MISMATCH")
        stamp = int(ledger._meta("lab_now_ms") or 0)
        cursor = int(ledger._meta("lab_cursor") or 0)
        start, end = phase["start_ms"], min(stamp, phase.get("end_ms") or stamp)
        accounts = ledger.portfolio_results()[mode]["rounds"]
        screens = {
            slug: json.loads(raw)
            for slug, raw in ledger.db.execute("SELECT slug,data FROM lab_rounds")
        }
        orders = {
            ident: json.loads(raw)
            for ident, raw in ledger.db.execute("SELECT id,data FROM intents")
        }
        fills: dict[str, list[tuple[dict[str, Any], dict[str, Any]]]] = defaultdict(list)
        for order_id, raw in ledger.db.execute("SELECT intent_id,data FROM fills"):
            order, fill = orders[order_id], json.loads(raw)
            fills[order["market"]["slug"]].append((order, fill))
        rows = []
        for slug in sorted(screens.keys() | accounts.keys()):
            t = int(slug.rsplit("-", 1)[1]) * 1000
            if t < start or t + ROUND_MS > end:
                continue  # ended markets only; partial current round cannot skew comparisons
            screen = screens.get(slug, {})
            account = accounts.get(slug, {})
            row = RoundResult(
                slug,
                t,
                account.get("realized_net_pnl", D(0)),
                account.get("fees", D(0)),
                bool(account.get("filled") or account.get("converted")),
                not account.get("unresolved", False),
                bool(screen.get("screens", 0)),
                screen.get("uncertain", True),
            )
            trades = sorted(fills[slug], key=lambda item: item[1]["timestamp_ms"])
            row.traded_shares = sum((D(f["quantity"]) for _, f in trades), D(0))
            buys = [(o, f) for o, f in trades if f["side"] == "BUY"]
            if buys and not account.get("converted"):
                order, fill = buys[0]
                # The first filled BUY is the entry condition, not an average of both
                # sides of an inventory strategy or an unfilled submitted limit.
                row.entry_price = D(fill["principal"]) / D(fill["quantity"])
                row.entry_seconds = (order["market"]["end_s"] * 1000 - fill["timestamp_ms"]) / 1000
                row.direction = "UP" if fill["token_id"] == order["market"]["up_token"] else "DOWN"
            exits = {o["reason"] for o, f in trades if f["side"] == "SELL"}
            row.exit_reason = ", ".join(sorted(exits)) if exits else None
            rows.append(row)
        return rows, {
            "as_of_ms": stamp,
            "cursor": cursor,
            "window_start_ms": start,
            "window_end_ms": end,
        }
    finally:
        ledger.close()


class ResearchReader:
    """Bounded selected-pair cache. No replay/writer, credential or network construction."""

    def __init__(self, path: Path) -> None:
        self.path = path
        self._cache: dict[tuple[str, ...], tuple[float, dict[str, Any]]] = {}
        self._lock = threading.Lock()

    @staticmethod
    def _document(path: Path) -> dict[str, Any]:
        with path.open("rb") as file:
            raw = file.read(16 * 1024 * 1024 + 1)
        if len(raw) > 16 * 1024 * 1024:
            raise ValueError("RESEARCH_DOCUMENT_TOO_LARGE")
        data = json.loads(raw)
        if not isinstance(data, dict):
            raise ValueError("RESEARCH_DOCUMENT_INVALID")
        return data

    def snapshot(
        self, suite: str, phase_id: str, selected: str, comparison: str, cohort: str
    ) -> dict[str, Any]:
        if suite not in ("directional", "order-flow") or cohort not in ("all", "unflagged"):
            raise ValueError("UNKNOWN_RESEARCH_FILTER")
        if not re.fullmatch(r"[a-zA-Z0-9_-]{1,64}", phase_id) or any(
            not re.fullmatch(r"[0-9a-f]{20}", i) for i in (selected, comparison)
        ):
            raise ValueError("UNKNOWN_RESEARCH_VARIANT")
        key = (suite, phase_id, selected, comparison, cohort)
        with self._lock:
            cached = self._cache.get(key)
            if cached and time.monotonic() - cached[0] < 20:
                return cached[1]
            root = self.path / ("lab" if suite == "directional" else "order-flow-lab")
            report = self._document(root / "report.json")
            registry = self._document(root / "study.json")
            if (
                report.get("environment") != "paper-lab"
                or registry.get("environment") != "paper-lab"
            ):
                raise ValueError("PAPER_RESEARCH_ONLY")
            phases = [p for p in report["phases"] if p["id"] == phase_id]
            if not phases:
                raise ValueError("UNKNOWN_RESEARCH_PHASE")
            phase = phases[0]
            catalog = {v["ident"]: v for v in phase["variants"]}
            registered = {v["ident"] for v in registry["variants"]}
            if any(i not in catalog or i not in registered for i in (selected, comparison)):
                raise ValueError("UNKNOWN_RESEARCH_VARIANT")
            samples, clocks = [], []
            for ident in (selected, comparison):
                journal = root / phase_id / ident / "ledger.sqlite"
                if not journal.resolve().is_relative_to(root.resolve()):
                    raise ValueError("RESEARCH_PATH_OUTSIDE_STUDY")
                rows, clock = read_rounds(
                    journal,
                    ident,
                    catalog[ident]["parameters"]["mode"],
                    phase,
                    registry["tape_identity"],
                )
                samples.append(rows)
                clocks.append(clock)
            end = min(c["window_end_ms"] for c in clocks)
            samples = [[r for r in rows if r.start_ms + ROUND_MS <= end] for rows in samples]
            a, b = (catalog[i] for i in (selected, comparison))
            differences = [
                {
                    "parameter": k,
                    "selected": a["parameters"].get(k),
                    "comparison": b["parameters"].get(k),
                }
                for k in sorted(a["parameters"].keys() | b["parameters"].keys())
                if a["parameters"].get(k) != b["parameters"].get(k)
            ]
            result = {
                "environment": "paper-research",
                "generated_ms": int(time.time() * 1000),
                "suite": suite,
                "phase": phase_id,
                "phase_kind": phase["kind"],
                "cohort": cohort,
                "trial_count": report["trial_count"],
                "clocks": clocks,
                "selected": {
                    "ident": selected,
                    "label": a["label"],
                    **summarize(samples[0], cohort=cohort),
                },
                "comparison": {
                    "ident": comparison,
                    "label": b["label"],
                    **summarize(samples[1], cohort=cohort),
                },
                "matched": compare_rounds(samples[0], samples[1], cohort=cohort),
                "parameter_differences": differences,
            }
            # No non-finite JSON or Decimal loss in financial values.
            result = json.loads(json.dumps(result, default=str, allow_nan=False))
            if len(self._cache) >= 8:
                self._cache.pop(next(iter(self._cache)))
            self._cache[key] = (time.monotonic(), result)
            return result

File: src/btc5m/dashboard_assets/research.js
"use strict";
let researchState=null, researchComparison=null, researchSelection=null, researchKey=null, researchLoaded=0, researchRequest=0, researchPending=null;
function researchOptions(id,items,value) {
  const el=$(id), signature=JSON.stringify(items);
  if(el.dataset.options!==signature){el.replaceChildren(...items.map(([v,label])=>{const opt=node("option","",label);opt.value=v;return opt;}));el.dataset.options=signature;}
  el.value=value;
}
function researchDefault(variants,selected) {
  const p=selected.parameters, opposite={reversal:"continuation",continuation:"reversal",absorption:"flow_continuation",flow_continuation:"absorption"}[p.signal];
  return (variants.find(v=>v.ident!==selected.ident&&v.parameters.signal===opposite&&v.parameters.lookback_seconds===p.lookback_seconds&&v.parameters.signal_threshold===p.signal_threshold)
    ||variants.find(v=>v.ident!==selected.ident&&v.family==="control"&&v.parameters.mode===p.mode)
    ||variants.find(v=>v.ident!==selected.ident&&v.family==="control")||selected).ident;
}
function renderResearchControls(data,phase) {
  const variants=phase.variants;
  if(!variants.length)return;
  if(!variants.some(v=>v.ident===labSelection))labSelection=variants[0].ident;
  const selected=variants.find(v=>v.ident===labSelection);
  if(researchSelection!==labSelection||!variants.some(v=>v.ident===researchComparison))researchComparison=researchDefault(variants,selected);
  researchSelection=labSelection;
  const options=variants.map(v=>[v.ident,v.label]);
  researchOptions("research-selected",options,labSelection);researchOptions("research-comparison",options,researchComparison);
  researchOptions("research-phase",data.phases.map(p=>[p.id,p.kind==="exploratory"?"Exploration":"Later-data test · "+dateTime(p.start_ms)]),labPhase);
  refreshResearch();
}
function refreshResearch(force=false) {
  if(view!=="lab"||labSection!=="research"||!labSelection||!researchComparison)return;
  const params=new URLSearchParams({suite:labSuite,phase:labPhase,selected:labSelection,comparison:researchComparison,cohort:$("research-cohort").value});
  const key=params.toString();
  if(!force&&key===researchKey&&(Date.now()-researchLoaded<15000||researchPending)){if(researchState)renderResearch();return;}
  if(researchPending)researchPending.abort();
  const request=++researchRequest, changed=key!==researchKey;
  researchKey=key;researchPending=new AbortController();
  const controller=researchPending, timeout=setTimeout(()=>controller.abort(new Error("Research request timed out. Retrying on the next refresh.")),15000);
  $("research-export").disabled=true;
  if(changed){researchState=null;$("research-error").hidden=true;$("research-title").textContent="Strategy comparison";$("research-conclusion").textContent="Reading the selected recorded results…";for(const id of ["research-scope","research-cost-note","research-match-note","research-metrics","research-round-chart","research-hour-chart","research-hours","research-outliers","research-costs","research-groups","research-matched","research-compare-chart","research-parameters","research-blocks","research-best","research-worst"])clear(id);}
  fetch("/api/research?"+key,{cache:"no-store",signal:researchPending.signal}).then(async response=>{
    if(!response.ok)throw Error("Research results are unavailable for this selection. The trading studies continue independently.");
    return response.json();
  }).then(data=>{
    if(request!==researchRequest)return;
    researchState=data;researchLoaded=Date.now();$("research-error").hidden=true;$("research-export").disabled=false;renderResearch();
  }).catch(error=>{
    if(request!==researchRequest||error.name==="AbortError")return;
    $("research-error").hidden=false;$("research-error").textContent=error.message;$("research-scope").textContent="Saved analysis may be stale; no trade results have been reset.";
  }).finally(()=>{clearTimeout(timeout);if(request===researchRequest)researchPending=null;});
}
function researchBars(id,points,note) {
  if(!points.length)return empty(clear(id),"No completed results in this cohort",note);
  const values=points.map(p=>Number(p.value)), stamps=points.map(p=>p.at_ms);
  const f=chartFrame(id,Math.min(0,...values),Math.max(0,...values),[stamps[0]-150000,stamps.at(-1)+150000],money);
  f.svg.append(svgNode("line",{x1:f.x(stamps[0]-150000),x2:f.x(stamps.at(-1)+150000),y1:f.y(0),y2:f.y(0),class:"lab-diagonal"}));
  const width=Math.min(20,Math.max(2,(f.x(stamps.at(-1)+150000)-f.x(stamps[0]-150000))/(points.length*1.5)));
  for(const p of points){const value=Number(p.value),y=f.y(value),zero=f.y(0);const rect=svgNode("rect",{x:f.x(p.at_ms)-width/2,y:Math.min(y,zero),width,height:Math.max(1,Math.abs(y-zero)),fill:value>=0?"#4cdbb4":"#f07882"});rect.append(svgNode("title",{},dateTime(p.at_ms)+" · "+money(p.value)+(p.flagged?" · flagged":"")));f.svg.append(rect);}
}
function researchComparisonChart(data) {
  const rows=data.matched.series;
  if(!rows.length)return empty(clear("research-compare-chart"),"No jointly comparable rounds","Missing and incomplete rounds are excluded from both sides.");
  let a=0,b=0;
  const points=rows.map(r=>[r.at_ms,a+=Number(r.selected),b+=Number(r.comparison)]),values=points.flatMap(p=>p.slice(1));
  const f=chartFrame("research-compare-chart",Math.min(0,...values),Math.max(0,...values),[points[0][0]-150000,points.at(-1)[0]+150000],money);
  for(const [col,color,label] of [[1,"#4cdbb4",data.selected.label],[2,"#7ba9f3",data.comparison.label]]){
    let path="",last=null;
    for(const p of points){path+=(last===null||p[0]-last>300000?"M":"L")+" "+f.x(p[0])+" "+f.y(p[col])+" ";last=p[0];const dot=svgNode("circle",{cx:f.x(p[0]),cy:f.y(p[col]),r:2,fill:color});dot.append(svgNode("title",{},dateTime(p[0])+" · "+label+" · "+money(p[col])));f.svg.append(dot);}
    const line=svgNode("path",{d:path,class:"plotline",stroke:color});line.append(svgNode("title",{},label));f.svg.append(line);
  }
}
function renderResearchGroup() {
  if(!researchState)return;
  const groups=researchState.selected.groups[$("research-group").value]||[];
  labTable("research-groups",["Condition","Completed rounds","Net profit","Mean / round","Win rate","Flagged rounds"],groups.map(g=>[g.label,count(g.rounds),money(g.net),money(g.mean),labPercent(g.win_rate),count(g.flagged_rounds)]));
}
function renderResearch() {
  if(!researchState||labSection!=="research")return;
  const d=researchState,a=d.selected,b=d.comparison,s=a.stats,c=a.counts,m=d.matched;
  $("research-title").textContent=a.label;
  const outlier=number(s.without_best_3);
  $("research-conclusion").textContent=!s.rounds?"No completed capital-used rounds match this cohort. "+(d.cohort==="unflagged"?"Try All recorded to inspect the excluded results.":"Choose another strategy, or wait for its held positions to finish."):money(s.net)+" net across "+s.rounds+" completed markets; the typical (median) result is "+money(s.median)+". "+(outlier===null?"There are too few results for the remove-three-winners check. ":"After removing up to three biggest winning rounds: "+money(outlier)+". "+(Number(s.net)>0&&outlier<=0?"The recorded profit depends on those wins. ":""))+s.flagged_rounds+" included rounds carry execution-data flags. These are exploratory comparisons, not a profitability verdict.";
  const cut=Math.min(...d.clocks.map(x=>x.window_end_ms));
  $("research-scope").textContent=(d.phase_kind==="holdout"?"REGISTERED LATER-DATA TEST":"EXPLORATION")+" · "+d.trial_count+" registered variants · ended markets from "+dateTime(d.clocks[0].window_start_ms)+" through "+dateTime(Math.floor(cut/300000)*300000)+" · "+(d.cohort==="all"?"includes flagged completed rounds":"unflagged subset; earlier trades still influenced cash and limits")+". "+c.incomplete_used+" incomplete capital-used rounds excluded; "+c.flagged_used+" recorded capital-used rounds flagged. Results can update as settlement completes. Analysis updated "+dateTime(d.generated_ms)+".";
  labMetrics("research-metrics",[
    ["NET PROFIT / COMPLETED ROUND",money(s.mean),s.rounds+" markets with capital used"],
    ["WIN RATE",labPercent(s.win_rate),s.wins+" wins · "+s.losses+" losses · "+s.flat+" flat"],
    ["AVERAGE WIN / LOSS",money(s.average_win)+" / "+money(s.average_loss),"After recorded fees"],
    ["PROFIT FACTOR",labNumber(s.profit_factor,2),"Winning dollars / losing dollars · undefined with no losses"],
  ]);
  researchBars("research-round-chart",a.rounds.map(r=>({at_ms:r.start_ms,value:r.net,flagged:r.uncertain})),"One observation per capital-used market.");
  researchBars("research-hour-chart",a.hourly.filter(h=>h.used_rounds).map(h=>({at_ms:h.at_ms,value:h.net,flagged:h.flagged_rounds>0})),"No hours with completed results.");
  labTable("research-hours",["Opening hour UTC","Net profit","Completed","Excluded capital-used","Observed slots / 12"],a.hourly.map(h=>[dateTime(h.at_ms),money(h.net),count(h.used_rounds),count(h.excluded_used),count(h.observed_slots)+" / 12"]));
  labTable("research-outliers",["Sample","Net profit"],[["All selected completed rounds",money(s.net)],["Without largest winning round",money(s.without_best_1)],["Without up to three largest wins",money(s.without_best_3)],["Largest win's share of all winning dollars",labPercent(s.best_win_share)],["Median completed round",money(s.median)]]);
  labTable("research-costs",["Cost assumption","Profit"],[["Before recorded fees",money(s.before_fees)],["After recorded fees ("+money(s.fees)+")",money(s.net)],...s.cost_stress.slice(1).map(x=>["Extra "+labNumber(Number(x.extra_per_share)*100,2)+"¢ per filled share",money(x.net)])]);
  $("research-cost-note").textContent="Additional costs apply to every bought and sold share, keeping the recorded fills and strategy unchanged. This does not simulate changed liquidity, queue priority, cash constraints or gas. "+(number(s.cost_headroom_per_share)!==null?"The recorded break-even extra cost is "+labNumber(Number(s.cost_headroom_per_share)*100,2)+"¢ per executed share.":"There is no measured positive extra-cost margin for this sample.");
  renderResearchGroup();
  labMetrics("research-matched",[
    ["SELECTED · SAME MARKETS",money(m.selected_net),a.label],
    ["COMPARISON · SAME MARKETS",money(m.comparison_net),b.label],
    ["NET DIFFERENCE",money(m.difference),m.matched_rounds+" matched market rounds"],
    ["OVERLAP WHEN EITHER TRADED",labPercent(m.common_used_fraction),m.both_used+" both used capital · "+m.selected_only+" selected only · "+m.comparison_only+" comparison only"],
  ]);
  researchComparisonChart(d);
  $("research-match-note").textContent="Green: "+a.label+"; blue: "+b.label+". Cumulative profit uses only the same comparable market rounds, including "+m.both_flat+" where both stayed flat. "+m.excluded_rounds+" unmatched, missing, flagged (if filtered), or unfinished rounds are excluded. Lines break at missing rounds. Profit correlation when either used capital: "+labNumber(m.active_round_correlation,2)+". Overlap/correlation describe shared evidence, not independent strategies or a combined portfolio.";
  labTable("research-parameters",["Displayed parameter difference","Selected","Comparison"],d.parameter_differences.map(p=>[human(p.parameter),String(p.selected??"—"),String(p.comparison??"—")]),"Displayed parameters match; check the full registered definitions before calling this a controlled ablation");
  const blocks=[...a.block_sensitivity.map(x=>({...x,series:"Selected"})),...m.block_sensitivity.map(x=>({...x,series:"Selected minus comparison"}))];
  labTable("research-blocks",["Quantity","Block size","Complete blocks","Included / available rounds","Mean / round","Descriptive 95% range"],blocks.map(x=>[x.series,x.minutes+" min",count(x.complete_blocks),x.included_rounds+" / "+x.available_rounds,money(x.mean_per_round),x.interval95?money(x.interval95[0])+" to "+money(x.interval95[1]):human(x.status)]));
  for(const [id,rows] of [["research-best",a.best_rounds.filter(r=>Number(r.net)>0)],["research-worst",a.worst_rounds.filter(r=>Number(r.net)<0)]])labTable(id,["Market opening UTC","Net profit","First entry","Flagged"],rows.map(r=>[dateTime(r.start_ms),money(r.net),quotePrice(r.entry_price),r.uncertain?"Yes":"No"]));
}
for(const button of document.querySelectorAll("[data-lab-section]"))button.addEventListener("click",()=>{
  labSection=button.dataset.labSection;
  for(const b of document.querySelectorAll("[data-lab-section]")){const selected=b===button;b.classList.toggle("selected",selected);b.setAttribute("aria-pressed",String(selected));}
  renderLab();
});
$("research-selected").addEventListener("change",()=>{labSelection=$("research-selected").value;renderLab();});
$("research-comparison").addEventListener("change",()=>{researchComparison=$("research-comparison").value;refreshResearch(true);});
$("research-phase").addEventListener("change",()=>{labPhase=$("research-phase").value;labSelection=null;renderLab();});
$("research-cohort").addEventListener("change",()=>refreshResearch(true));
$("research-group").addEventListener("change",renderResearchGroup);
$("research-export").addEventListener("click",()=>{
  if(!researchState)return;
  const url=URL.createObjectURL(new Blob([JSON.stringify(researchState,null,2)],{type:"application/json"}));
  const link=node("a");link.href=url;link.download="btc5m-strategy-research.json";link.click();setTimeout(()=>URL.revokeObjectURL(url),1000);
});
window.addEventListener("resize",()=>{if(view==="lab")renderResearch();});
